src/manifolds/StiefelSubmersionMetric.jl

```diff
@@ -0,0 +1,293 @@
+
+The exponential map is given by
+````math
+\exp_p X = \operatorname{Exp}\bigl(
+    -\frac{2α+1}{α+1} p p^\mathrm{T} X p^\mathrm{T} +
+    X p^\mathrm{T} - p X^\mathrm{T}
+\bigr) p \operatorname{Exp}\bigl(\frac{\alpha}{\alpha+1} p^\mathrm{T} X\bigr)
+````
+"""
+exp(::MetricManifold{ℝ,Stiefel{n,k,ℝ},<:StiefelSubmersionMetric}, ::Any...) where {n,k}
+
+function exp!(
+    M::MetricManifold{ℝ,Stiefel{n,k,ℝ},<:StiefelSubmersionMetric},
+    q,
+    p,
+    X,
+) where {n,k}
+    α = metric(M).α
+    T = Base.promote_eltype(q, p, X)
+    # TODO:
+    # - dispatch to exp! for α = -1/2 and α = 0 for efficiency
+    # - handle rank-deficient QB
+    if k ≤ div(n, 2)
+        # eq. 11
+        A = allocate(q, T, Size(k, k))
+        Y = allocate(X, T)
+        F = allocate(A, T, Size(2k, 2k))
+        G = allocate(p, T, Size(n, 2k))
+        mul!(A, p', X)
+        copyto!(Y, X)
+        mul!(Y, p, A, -1, true)
+        Q, B = qr(Y)
+        @views begin
+            copyto!(G[1:n, 1:k], p)
+            copyto!(G[1:n, (k + 1):(2k)], Matrix(Q))
+            F[1:k, 1:k] .= A ./ (α + 1)
+            F[1:k, (k + 1):(2k)] .= -B'
+            F[(k + 1):(2k), 1:k] .= B
+            fill!(F[(k + 1):(2k), (k + 1):(2k)], false)
+        end
+        rmul!(A, α / (α + 1))
+        mul!(q, G, @views(exp(F)[1:(2k), 1:k]) * exp(A))
+    elseif n == k
+        A = allocate(q, T)
+        C = allocate(q, T)
+        mul!(A, p', X, α / (α + 1), false)
+        mul!(C, X, p', inv(α + 1), false)
+        expC = exp(C)
+        tmp = mul!(C, p, exp(A))
+        mul!(q, expC, tmp)
+    else  # n/2 < k < n
+        # eq. 8
+        A = p' * X
+        C = allocate(q, T, Size(n, n))
+        tmp = allocate(q, T)
+        mul!(C, X, p')
+        C .-= C'
+        mul!(tmp, p, A)
+        mul!(C, tmp, p', -(2α + 1) / (α + 1), true)
+        rmul!(A, α / (α + 1))
+        mul!(tmp, p, exp(A))
+        mul!(q, exp(C), tmp)
+    end
+    return q
+end
+
+@doc raw"""
+    inner(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, <:StiefelSubmersionMetric}, p, X, Y)
+
+Compute the inner product on the [`Stiefel`](@ref) manifold with respect to the
+[`StiefelSubmersionMetric`](@ref). The formula reads
+```math
+g_p(X,Y) = \operatorname{tr}\bigl( X^{\mathrm{T}}(I_n - \frac{2α+1}{2(α+1)}pp^{\mathrm{T}})Y \bigr),
+```
+where ``α`` is the parameter of the metric.
+"""
+function inner(
+    M::MetricManifold{ℝ,Stiefel{n,k,ℝ},<:StiefelSubmersionMetric},
+    p,
+    X,
+    Y,
+) where {n,k}
+    α = metric(M).α
+    T = typeof(one(Base.promote_eltypeof(p, X, Y, α)))
+    if n == k
+        return T(dot(X, Y)) / (2 * (α + 1))
+    elseif α == -1 // 2
+        return T(dot(X, Y))
+    else
+        return dot(X, Y) - (T(dot(p'X, p'Y)) * (2α + 1)) / (2 * (α + 1))
+    end
+end
+
+function inverse_retract!(
+    M::MetricManifold{ℝ,Stiefel{n,k,ℝ},<:StiefelSubmersionMetric},
+    X,
+    p,
+    q,
+    method::ShootingInverseRetraction{ExponentialRetraction},
```

Comment:
If we defined specialized point and vector types that represented a point and its tangent vectors on `Stiefel` in the `k × k` subspace, then perhaps by implementing the necessary `exp`, `inner`, `project`, and addition methods for these types, we could delete this method entirely. Might be overkill though, since we don't otherwise have a use for these. Perhaps for a future PR if they seem to be useful.

Reply:
I have now documented the types so it is hopefully clear.